File: prog3/trab_java/utilitarios/capes/PPG.java
package utilitarios.capes;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import utilitarios.capes.producoes.*;

/**
 * Classe comparadora que realiza a comparacao entre dois PPGs pelo nome
 */
class ComparadorNome implements Comparator<PPG>{
    public int compare(PPG a, PPG b){
        return a.getNome().compareTo(b.getNome());
    }
}

/**
 * Classe que representa um Programa de Pos Graducao.
 */
public class PPG implements Comparable<PPG>{
    private String nome;
    private ArrayList<IES> instituicoes;
    private String codigo;
    private ArrayList<Producao> producoes;
    private int qtdProducoes = 0;
    private int qtdPags = 0;
    private boolean rede = false;
    private int qtdAnal = 0;
    private int qtdArtjr = 0;
    private int qtdArtpe = 0;
    private int qtdLivro = 0;
    private int qtdPartmu = 0;
    private int qtdTradu = 0;
    private int qtdOutro = 0;


    PPG(String nome, String codigo){
        this.nome = nome;
        this.instituicoes = new ArrayList<>();
        this.codigo = codigo;
        this.producoes = new ArrayList<Producao>();
    }

    @Override
    public int compareTo(PPG o) {
        return this.codigo.compareTo(o.getCodigo());
    }

    @Override
    public boolean equals(Object o){
        PPG ppg = (PPG) o;
        return this.codigo.equals(ppg.getCodigo());
    }
    /**
     * Metodo que cadastra uma producao {@code}producao{@code} ao PPG
     * @param producao a ser cadastrada
     */
    public void addProducao(Producao producao){
        if(!this.producoes.contains(producao))this.producoes.add(producao);
        this.qtdPags += producao.getQtdPags();
        this.qtdProducoes++;

        if(producao instanceof Anal) this.qtdAnal++;
        if(producao instanceof ArtigoJR) this.qtdArtjr++;
        if(producao instanceof ArtigoPE) this.qtdArtpe++;
        if(producao instanceof Livro) this.qtdLivro++;
        if(producao instanceof Partitura) this.qtdPartmu++;
        if(producao instanceof Traducao) this.qtdTradu++;
        if(producao instanceof Outro) this.qtdOutro++;

    }

    /**
     * Metodo que verifica se o PPG possui uma instituicao {@code}ies{@code} e, caso nao possua, cadastra a mesma ao PPG. 
     * Se o PPG possuir mais de uma instituicao cadastrada, o metodo qualifica ele como Rede.
     * @param ies a ser checada
     */
    public void checaIES(IES ies){
        if(!instituicoes.contains(ies))
            instituicoes.add(ies);

        if(instituicoes.size() > 1) this.rede = true;
        Collections.sort(this.instituicoes);    
    }

    /**
     * Metodo que imprime as instituicoes cadastradas do PPG.
     */
    public void imprimeRede(){
        System.out.println("Programa: " + this.getNome());
        System.out.println("Instituicoes:");
        for(IES iter : this.instituicoes){
            System.out.print("\t");
            System.out.println("- " + iter.getSigla() + " (" + iter.getNome() + ")");
        }
    }

    /**
     * Metodo que imprime a quantidade de cada subtipo de producao o PPG possui.
     */
    public void imprimeProducoes(){
        System.out.println("Quantidade de producoes por tipo:");
        System.out.println("\t- Artigos em anais de eventos: " + this.qtdAnal);
        System.out.println("\t- Artigos em jornais e revistas: " + this.qtdArtjr);
        System.out.println("\t- Artigos em periodicos cientificos: " + this.qtdArtpe);
        System.out.println("\t- Livros: " + this.qtdLivro);
        System.out.println("\t- Partituras musicais: " + this.qtdPartmu);
        System.out.println("\t- Traducoes: " + this.qtdTradu);
        System.out.println("\t- Outros: " + this.qtdOutro);
        System.out.println("\nTotal de paginas produzidas pelo PPG: " + this.qtdPags);

    }

    /**
     * Metodo que imprime informacoes de forma simplificada sobre o PPG
     */
    public void imprimeSimples(){
        System.out.println("\t- " + this.nome + ": " + this.qtdProducoes + " producoes");
    }

    private void imprimeCabecalho(String tipo){
        if(tipo.equals("Anal")) Anal.imprimeCabecalho();
        if(tipo.equals("ArtigoPE")) ArtigoPE.imprimeCabecalho();
        if(tipo.equals("ArtigoJR")) ArtigoJR.imprimeCabecalho();
        if(tipo.equals("Livro")) Livro.imprimeCabecalho();
        if(tipo.equals("Partitura")) Partitura.imprimeCabecalho();
        if(tipo.equals("Traducao")) Traducao.imprimeCabecalho();
        if(tipo.equals("Outro")) Outro.imprimeCabecalho();
    }
    /**
     * 
     * Metodo que imprime as informacoes de um tipo de producao especifico do PPG.
     * O parametro fornecido deve ser: {@code}Anal{@code} {@code}ArtigoPE{@code} {@code}ArtigoJR{@code} {@code}Livro{@code} {@code}Partitura{@code} {@code}Traducao{@code} {@code}Outro{@code}.
     * @param tipo o tipo da producao que se deseja observar.
     */
    public void imprimeCsv(String tipo){
        imprimeCabecalho(tipo);
        List prods = new ArrayList<>();
        
        for(Producao iter : this.producoes){
            try {
                if(Class.forName(tipo).isInstance(iter)) prods.add(iter);
            } catch (ClassNotFoundException e){
                continue;
            }
        }
        Collections.sort(prods);
        for(Object iter : prods) System.out.println(iter.toString());
    }

    public String getNome() {
        return this.nome;
    }

    public String getCodigo() {
        return this.codigo;
    }

    public int getQtdPags(){
        return this.qtdPags;
    }

    public boolean getRede(){
        return this.rede;
    }
    
}

File: prog3/trab_cpp/PPG.cpp
#include "PPG.hpp"
#include <iostream>
#include "StringUtils.h"

/* Construtor da classe ppg_exception, a excessao padrao dos ppgs.*/
ppg_exception::ppg_exception(string msg){
    this->msg = msg;
}

const char* ppg_exception::what() const throw(){
    return msg.c_str();
}

/* Excessao de tipo invalido. */
tipo_invalido_exception::tipo_invalido_exception(): ppg_exception("Tipo invalido."){}

PPG::PPG(string nome, string codigo){
    this->nome = nome;
    this->codigo = codigo;
}

PPG::~PPG(){
    for(auto iter : producoes) delete iter;
}

string PPG::getCodigo(){
    return codigo;
}

string PPG::getNome(){
    return nome;
}

int PPG::getQtdProd(int subtipo){
    int counter = 0;
    for(auto iter : producoes) if(iter->getSubtipo() == subtipo) counter++;
    return counter;
}

int PPG::getQtdProd(){
    return producoes.size();
}

/* Metodo que retorna a quantidade de paginas produzido pelo PPG. */
int PPG::getQtdPags(){
    int counter = 0;
    //Soma a quantidade de paginas de cada producao associada ao PPG.
    for(auto iter : producoes) counter += iter->getQtdPags();
    return counter;
}

/* Metodo que adiciona uma IES associada ao PPG. Checa antes se a IES ja foi cadastrada ao PPG. */
void PPG::addInstituicao(IES* ies){
    string chave = ies->getSigla() + ies->getNome();
    if(instituicoes.find(chave) == instituicoes.end()) instituicoes.insert({chave, ies});
}

/* Metodo que insere uma nova producao no PPG. */
void PPG::addProducao(Producao* prod){
    producoes.push_back(prod);
}

/* Metodo que retorna se o PPG eh rede ou nao. */
bool PPG::ehRede(){
    return instituicoes.size() > 1;
}

//Funcao auxiliar para a ordenacao das instituicoes.
bool _ordena_Nome(IES* ies1, IES* ies2){
    return *ies1 < *ies2;
}

/* Metodo que imprime as informacoes da rede de instituicoes relacionado ao PPG. */
void PPG::imprimeRede(){
    //Criando um vector temporario para futura ordenacao
    vector<IES*> temp;

    //Inserindo no vector as instituicoes do PPG
    for(auto iter : instituicoes) temp.push_back(iter.second);
    sort(temp.begin(), temp.end(), _ordena_Nome);

    //Imprimindo as informacoes de cada instituicao
    for(IES* iter : temp) cout << "\t- " << *iter << endl;
}

//Funcao auxiliar que transforma o subtipo da producao em um valor numerico.
int geraCodigo(string producao){
    if(producao == "anais") return 8;
    if(producao == "artjr") return 9;
    if(producao == "artpe") return 25; 
    if(producao == "livro") return 26;
    if(producao == "partmu") return 28;
    if(producao == "tradu") return 21;
    if(producao == "outro") return 10;
    return -1;
}

//Funcao auxiliar para a ordenacao de producoes
bool _ordena_producao(Producao* p1, Producao* p2){
    return *p1 < *p2;
}

//Funcao auxiliar para a insersao de producoes no vector auxiliar para imprimir o csv
bool contem(vector<Producao*> temp, Producao* prod){
    for(auto iter : temp) if(*prod == *iter) return true;
    return false;
}

/* Metodo que gera e imprime as informacoes de uma producao referente a um PPG em formato csv.
Caso o tipo de producao que se deseja imprimir nao existe, lanca "Tipo invalido." */
void PPG::imprimeCsv(string producao){

    //checando se o subtipo da producao existe
    int subtipo = geraCodigo(producao);
    if(subtipo == -1){
        throw tipo_invalido_exception();
    }

    //criando o vector auxiliar e inserindo apenas as producoes desejadas.
    vector<Producao*> temp;
    for(auto iter : producoes) if(iter->getSubtipo() == subtipo && !contem(temp, iter)) temp.push_back(iter);
    sort(temp.begin(), temp.end(), _ordena_producao);

    //imprimindo o cabecalho da producao desejada
    cout << temp[0]->cabecalho() << endl;
    for(auto iter : temp) cout << *iter << endl;
}

/* Metodo que imprime as informacoes estatisticas do PPG da seguinte forma:
 "Instituicoes:
<instituicoes_relacionadas_ao_ppg>
Quantidade de producoes por tipo:
Artigos em anais de eventos: <qtd_prod_anal>
Artigos em jornais e revistas: <qtd_prod_artjr>
Artigos em periodicos cientificos: <qtd_prod_artpe>
Livros: <qtd_prod_livro>
Partituras musicais: <qtd_prod_partmu>
Traducoes: <qtd_prod_tradu>
Outros: <qtd_prod_outro>
Total de paginas produzidas pelo PPG: <total_pags_ppg>"*/
void PPG::imprimeInfo(){
    cout << "Instituicoes:" << endl;
    imprimeRede();
    cout << "\nQuantidade de producoes por tipo:" << endl;
    cout << "\t- Artigos em anais de eventos: " << getQtdProd(8) << endl;
    cout << "\t- Artigos em jornais e revistas: " << getQtdProd(9) << endl;
    cout << "\t- Artigos em periodicos cientificos: " << getQtdProd(25) << endl;
    cout << "\t- Livros: " << getQtdProd(26) << endl;
    cout << "\t- Partituras musicais: " << getQtdProd(28) << endl;
    cout << "\t- Traducoes: " << getQtdProd(21) << endl;
    cout << "\t- Outros: " << getQtdProd(10) << endl;
    cout << "\nTotal de paginas produzidas pelo PPG: " << getQtdPags() << endl;


}

//sobrescrita do operador << para o PPG.
ostream& operator << (ostream& out, PPG& ppg){
    out << ppg.getCodigo() << ": " << ppg.getNome();
    return out;
}